### PR TITLE
Initial qrCode feature Implementatiaon

### PR DESCRIPTION
**Pull request by stanley92**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -26,5 +38,11 @@
   "author": "",
   "bugs": {
     "url": "https://github.com/AdactiveSAS/qrcode.js/issues"
+  },
+  "files": [
+    "build"
+  ],
+  "dependencies": {
+    "qrcode-generator": "^1.4.0"
   }
```
Comment (moroine) [inline package.json]: Missing package.json keys:
- https://docs.npmjs.com/files/package.json#keywords
- https://docs.npmjs.com/files/package.json#license
- https://docs.npmjs.com/files/package.json#people-fields-author-contributors
- https://docs.npmjs.com/files/package.json#bugs
-

File: package.json [hunk 2]
```diff
@@ -1,23 +1,35 @@
...
+    "rollup-plugin-babel": "^2.7.1",
+    "rollup-plugin-commonjs": "^8.0.2",
+    "rollup-plugin-node-resolve": "^3.0.0"
   },
   "scripts": {
-    "eslint": "eslint src"
+    "build": "rollup -c",
+    "dev": "rollup -c -w",
+    "test": "node test/test.js",
+    "pretest": "npm run build"
```
Comment (moroine) [inline package.json]: Why did you remove eslint ?

Comment (moroine) [reply package.json]: You got test, you must add continuous integration ;)

File: src/qrcode.js [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+    const w = ctx.measureText(settings.label).width;
+    const sh = settings.mSize;
+    const sw = w / size;
+    const sl = (1 - sw) * settings.mPosX;
+    const st = (1 - sh) * settings.mPosY;
+    const sr = sl + sw;
+    const sb = st + sh;
+    const pad = 0.01;
+
+    if (settings.mode === 1) {
```
Comment (moroine) [inline src/qrcode.js]: Use your constant here

File: src/qrcode.js [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+}
+
+function drawBackground(qr, context, settings) {
+    if (settings.background) {
+        context.fillStyle = settings.background;
+        context.fillRect(settings.left, settings.top, settings.size, settings.size);
+    }
+
+    const { mode } = settings;
+    if (mode === 1 || mode === 2) {
```
Comment (moroine) [inline src/qrcode.js]: Same use constant here

File: src/qrcode.js [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+        drawBackgroundLabel(qr, context, settings);
+    } else if (mode === 3 || mode === 4) {
```
Comment (moroine) [inline src/qrcode.js]: And here

File: src/qrcode.js [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+
+function createMinQRCode(text, level, minVersion, maxVersion, quiet) {
+    minVersion = Math.max(1, minVersion || 1);
+    maxVersion = Math.min(40, maxVersion || 40);
+    for (let version = minVersion; version <= maxVersion; version += 1) {
+        try {
+            return createQRCode(text, level, version, quiet);
+        } catch (err) { /* empty */ }
+    }
+    return undefined;
```
Comment (moroine) [inline src/qrcode.js]: If you reach this, you should effectively throw an Error, like too much data

File: src/qrcode.js [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+
+    imageElement.addEventListener('load', () => drawOnCanvas(element, options, imageElement));
+    imageElement.addEventListener('error', () => {
+        throw new Error('Please provide a valid image');
+    });
+
+    if (image && typeof image === 'string') {
+        imageElement.src = image;
+    } else {
+        throw new Error('An Image is required for Mode 3 and Mode 4');
```
Comment (moroine) [inline src/qrcode.js]: This is not an image which is required but an URL !

File: src/qrcode.js [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+    return element;
+}
+
+function loadImage(element, options) {
+    const { image } = options;
+    const imageElement = new Image();
+
+    imageElement.addEventListener('load', () => drawOnCanvas(element, options, imageElement));
+    imageElement.addEventListener('error', () => {
+        throw new Error('Please provide a valid image');
```
Comment (moroine) [inline src/qrcode.js]: console log the error from image in order to let the integrator know what is the source of the issue

File: src/qrcode.js [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+
+    if (image && typeof image === 'string') {
+        imageElement.src = image;
+    } else {
+        throw new Error('An Image is required for Mode 3 and Mode 4');
+    }
+}
+
+function createCanvas(element, options) {
+    if (options.mode === 3 || options.mode === 4) {
```
Comment (moroine) [inline src/qrcode.js]: Use constant

File: src/qrcode.js (hunk 7 above)
Comment (moroine) [reply src/qrcode.js]: And use constant names

File: src/qrcode.js [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+    constructor(ele) {
+        this.element = ele;
+    }
+
+    /**
+     * Returns a Promise resolved when done with the canvas element
+     * @returns {*}
+     */
+    generate(opts) {
+        if (!(this.element.tagName === 'DIV' || this.element.tagName === 'CANVAS')) {
```
Comment (moroine) [inline src/qrcode.js]: Better with 

```javascript
if ( !( this.element instanceof HTMLDivElement || this.element instanceof HTMLCanvasElement) )
```

File: src/qrcode.js [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+            console.log(`[You provided an ${this.element.tagName} element]`);
```
Comment (moroine) [inline src/qrcode.js]: That console log is not need IMO

File: src/qrcode.js (hunk 10 above)
Comment (moroine) [reply src/qrcode.js]: And what happen if the element has no property tagName (if it's undefined ?)

Comment (moroine) [reply src/qrcode.js]: You should move that test in constructor and throw TypeError instead

File: src/qrcode.js [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+        }
+
+        const options = Object.assign(defaultOptions, opts);
+
+        if (this.element && this.element.tagName === 'DIV') {
+            const newCanvasElement = document.createElement('canvas');
+            createCanvas(newCanvasElement, options);
+            this.element.appendChild(newCanvasElement);
+        } else if (this.element && this.element.tagName === 'CANVAS') {
+            this.element = createCanvas(this.element, options);
```
Comment (moroine) [inline src/qrcode.js]: `createCanvas` is not a proper name

Comment (moroine) [reply src/qrcode.js]: So what happen with a div element and multiple generate ? This will append new canvas each time ? 

I think you should refresh the first canvas

**Review by moroine — CHANGES_REQUESTED** [review]
Good job !

You forget to add README.md with small doc and example (using gh-pages for example)

File: src/qrcode.js (hunk 10 above)
Comment (stanley92) [reply src/qrcode.js]: Will `throw new TypeError(message)` be enough? Or should I do - `throw new TypeError(message, file, lineNumber)` as per MDN documentation?

File: README.md [hunk 13]
```diff
@@ -1,2 +1,174 @@
...
+OR
+```javascript
+    yarn add qrcode.es
+```
+
+## Usage
+
+### ES6 Modules
+```javascript
+import QrCode, {MODES} from 'qrcode.es';
```
Comment (moroine) [inline README.md]: I don't think so 
 !

https://github.com/AdactiveSAS/qrcode.js/blob/f4baf58c05a5d24b9a182a2008a0a249c40a6286/src/index.js

You didn't test it ?!

File: README.md [hunk 14]
```diff
@@ -1,2 +1,174 @@
...
+    
+const qrCodeSetting = {
+    size: 400,
+    minVersion: 8,
+    text: 'http://adactive.com/',
```
Comment (moroine) [inline README.md]: Use https link :wink:

File: README.md [hunk 15]
```diff
@@ -1,2 +1,174 @@
...
+                    console.log("[IMAGE URL] : ", qrCodeImageUrl);
+                });
+
+        </script>
+    </body>
+</html>
+
+```
+
+## Functions
```
Comment (moroine) [inline README.md]: Should be API section and the generate signature is not true and miss the constructor

File: README.md [hunk 16]
```diff
@@ -1,2 +1,174 @@
...
+fucntion generate(ele: HTMLCanvasElement | HTMLDivElement): Promise
+function getImage(): string
+```
+
+## Options
+
+| Name       | Type   | Default    | Description                                                     |
+| ---------- | ------ | ---------- | --------------------------------------------------------------- |
+| minVersion | number | 1          | Version Range: Minimum 1                                        |
+| maxVersion | number | 40         | Version Range: Maximum 40                                       |
```
Comment (moroine) [inline README.md]: Version range of what ?

File: README.md [hunk 17]
```diff
@@ -1,2 +1,174 @@
...
+| ecLevel    | string | 'H'        | Error Correction Level                                          |
```
Comment (moroine) [inline README.md]: What are the consequencies?

File: README.md [hunk 18]
```diff
@@ -1,2 +1,174 @@
...
+| left       | number | 0          | Offset in pixel on existing Canvas Element                      |
+| top        | number | 0          | Offset in pixel on existing Canvas Element                      |
```
Comment (moroine) [inline README.md]: Why needs top and left ? Just adapt the canvas to the size

File: README.md [hunk 19]
```diff
@@ -1,2 +1,174 @@
...
+| size       | number | 500        | Size of QrCode in pixel                                         |
+| fill       | string | '#000'     | Color of Module in Color Code                                   |
```
Comment (moroine) [inline README.md]: What is module ? Maybe need a gloassary

File: README.md [hunk 20]
```diff
@@ -1,2 +1,174 @@
...
+| background | string | null       | Background color or `null` for transparent elements             |
```
Comment (moroine) [inline README.md]: for transparent elements ?

File: README.md [hunk 21]
```diff
@@ -1,2 +1,174 @@
...
+| text       | string | 'no text'  | Content of QrCode                                               |
```
Comment (moroine) [inline README.md]: should not be an options IMO

File: README.md [hunk 22]
```diff
@@ -1,2 +1,174 @@
...
+const MODES = {
+    NORMAL: 0,
+    DRAW_WITH_LABEL_STRIP: 1,
+    DRAW_WITH_LABEL_BOX: 2,
+    DRAW_WITH_IMAGE_STRIP: 3,
+    DRAW_WITH_IMAGE_BOX: 4,
+};
+
+//Default Options
+const options = {
```
Comment (moroine) [inline README.md]: Don't care

File: README.md [hunk 23]
```diff
@@ -1,2 +1,174 @@
...
+| mPosX      | number | 0.5        | Position of Label or Image on x-Axis                            |
+| mPosY      | number | 0.5        | Position of Label or Image on y-Axis                            |
+| fontname   | string | 'sans'     | Font Name to be used                                            |
+| fontcolor  | string | '#000'     | Font Colour in Color Code                                       |
+| label      | string | 'no label' | Label used for Mode with Label                                  |
+| image      | string | null       | Image used for Mode with Image                                  | 
+
+```javascript
+//Modes available for drawing of QrCode
+const MODES = {
```
Comment (moroine) [inline README.md]: Should be a table

File: package.json [hunk 24]
```diff
@@ -1,30 +1,62 @@
 {
-  "name": "qrcode.js",
-  "version": "0.0.0",
+  "name": "qrcode.es",
   "description": "QR Code Generator",
+  "version": "0.0.1-rc.5",
+  "main": "build/qrcode.cjs.js",
```
Comment (moroine) [inline package.json]: doesn't exists

**Review by moroine — CHANGES_REQUESTED** [review]

**Review by antoinebigard — APPROVED** [review]

**Review by moroine — APPROVED** [review]
